riscv: Allow SOC to override arch_irq_{lock,unlock,unlocked}

carlocaione opened the pull request:
RISC-V has a modular design. Some hardware with a custom interrupt controller needs a bit more work to lock / unlock IRQs.
    
Account for this hardware by introducing a set of new z_soc_irq_* functions that can override the default behaviour.
    
Signed-off-by: Carlo Caione <ccaione@baylibre.com>

gmarull commented:
> Account for this hardware by introducing a set of new `SOC_RISCV_ARCH_IRQ_*` macros that can override the default behaviour. The natural place to specialize these macros should be the `soc.h` header.

Note: `soc.h` is today a can of worms, specially on ARM. Some platforms don't even have it (as it is in most cases not needed, or there are better alternatives). I think we should have specific purpose headers, as we do for pinctrl (pinctrl_soc.h), your recent PR for RISCV ISR stuff, etc.

carlocaione commented:
> Note: `soc.h` is today a can of worms, specially on ARM. Some platforms don't even have it (as it is in most cases not needed, or there are better alternatives). I think we should have specific purpose headers, as we do for pinctrl (pinctrl_soc.h), your recent PR for RISCV ISR stuff, etc.

`soc.h` is kinda a natural choice for this because of https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/riscv/arch.h#L29

I can leave that comment out.

gmarull commented:
> > Note: `soc.h` is today a can of worms, specially on ARM. Some platforms don't even have it (as it is in most cases not needed, or there are better alternatives). I think we should have specific purpose headers, as we do for pinctrl (pinctrl_soc.h), your recent PR for RISCV ISR stuff, etc.
> 
> `soc.h` is kinda a natural choice for this because of https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/arch/riscv/arch.h#L29
> 
> I can leave that comment out.

Yep, it is now, but soc.h comes with many problems since its purpose is not clear and propagates includes everywhere, leading to all sorts of problems. Let's leave it as is now, but find better alternatives in the future.

stephanosio commented on include/zephyr/arch/riscv/arch.h:
```diff
@@ -220,6 +220,7 @@ extern void z_irq_spurious(const void *unused);
  */
 static ALWAYS_INLINE unsigned int arch_irq_lock(void)
 {
+#ifndef SOC_RISCV_ARCH_IRQ_LOCK
```
What would be a reason to not follow the existing `z_soc_irq_...` pattern here?

https://github.com/zephyrproject-rtos/zephyr/blob/abe50cc92991389e2508de8ea721bee36cd5fb17/include/zephyr/arch/arm/aarch32/irq.h#L54-L57

https://github.com/zephyrproject-rtos/zephyr/blob/abe50cc92991389e2508de8ea721bee36cd5fb17/include/zephyr/arch/xtensa/irq.h#L77-L80

https://github.com/zephyrproject-rtos/zephyr/blob/abe50cc92991389e2508de8ea721bee36cd5fb17/include/zephyr/arch/arm64/irq.h#L53-L56

carlocaione replied:
Because vendors can do whatever they want with RISC-V and we have to support it. So in the use-case I have in mind the core is shipping a PLIC controller following the spec but a new CSR was added to the ISA that is used to implement critical sections (thus used in the `arch_irq_*` family)

stephanosio replied:
I am still not seeing why the function should not be named `z_soc_irq_*` (e.g. `z_soc_irq_lock`), that can be optionally "mapped" to `arch_irq_*` when the relevant Kconfig is enabled (e.g. `CONFIG_RISCV_CUSTOM_IRQ_CSR`).

Relying on `#ifndef SOC_RISCV_ARCH_IRQ_*`, instead of a Kconfig to explicitly control this behaviour is a rather fragile approach.

Also it looks like the return value of `SOC_RISCV_ARCH_IRQ_LOCK` is not properly handled?

carlocaione replied:
@stephanosio pushed a new version based on your suggestions

carlocaione committed:
riscv: Allow SOC to override arch_irq_{lock,unlock,unlocked}

RISC-V has a modular design. Some hardware with a custom interrupt
controller needs a bit more work to lock / unlock IRQs.

Account for this hardware by introducing a set of new
z_soc_irq_* functions that can override the default behaviour.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented on include/zephyr/arch/riscv/arch.h:
```diff
@@ -236,15 +240,23 @@ static ALWAYS_INLINE unsigned int arch_irq_lock(void)
+	return z_soc_irq_unlocked(key);
+#else
 	return (key & MSTATUS_IEN) != 0;
+#endif
```
because in that case you will get a compiler warning like:

```
include/zephyr/sys/arch_interface.h:247:28: warning: 'arch_irq_lock' declared 'static' but never defined [-Wunused-function]
```

Because these functions (differently from what happens with `arch_irq_enable`) are declared as `static inline` .

edersondisouza replied:
Oh, fair.

nashif commented on arch/riscv/Kconfig:
```diff
@@ -46,6 +46,18 @@ config INCLUDE_RESET_VECTOR
 	  Include the reset vector stub, which initializes the stack and
 	  prepares for running C code.
 
+config RISCV_SOC_HAS_CUSTOM_IRQ_LOCK_OPS
```
why not make it generic in naming, similar to many others ..HAS_CUSTOM_... kconfigs? Eventually yhis could be used by other architectures, i.e. we have
- ARCH_HAS_CUSTOM_BUSY_WAIT
- CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS
- ARCH_HAS_CUSTOM_SWAP_TO_MAIN
.....

carlocaione replied:
Just to be clear, you are suggesting to make these options available for all the architectures and possibly move this to `arch/Kconfig` ?

nashif commented on include/zephyr/arch/riscv/arch.h:
```diff
@@ -220,6 +220,9 @@ extern void z_irq_spurious(const void *unused);
 static ALWAYS_INLINE unsigned int arch_irq_lock(void)
 {
+#ifdef CONFIG_RISCV_SOC_HAS_CUSTOM_IRQ_LOCK_OPS
+	return z_soc_irq_lock();
```
i do not see z_soc_irq_lock defined anywhere

carlocaione replied:
this is kinda on purpose. You want to keep the inlining so it is implied to have these functions declared as `static inline` directly in some header file (in my tests I have those in `soc.h` for example).

carlescufi commented:
@fkokosinski @kgugala ping

gmarull commented:
@carlocaione please fix conflicts
